### PR TITLE
C51-291: civicase popover

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview

This PR allows the contact popover to remain open when moving the mouse from the contact icon to the popover card.

It also fixes the style for the additional contacts popover since it was not the same as the one in the invision design.

## Before
pr-before

## After
pr-after

### Civicase Popover

A new civicase popover component was created for the following reasons:

* UIB Popover only supports basic hover of the original toggle element. There's no built in support to keep the popover visible when moving from the toggle element to the popover card.
* UIB Popover can only be appended to the body, you can't append it to an arbitrary element. The popover would not be properly rendered at the body, only when inside the `#bootstrap-theme` container element. We need to append it to this element when the popover is for an element that is inside an `overflow: scroll` or similar.
* UIB Popover can't be positioned on top of an arbitrary coordinate/element. If you try to manually position it elsewhere, it will keep being repositioned back to its original toggle element. For the [Activities Calendar](https://github.com/compucorp/uk.co.compucorp.civicase/blob/develop/ang/civicase/ActivitiesCalendar.js) we need the popover to open on top of a specific calendar day, which is not possible with UIB Popover.

Also, there are these other benefits to having this new component:

* We can remove the [Popover Append to Element](https://github.com/compucorp/uk.co.compucorp.civicase/blob/develop/ang/civicase/PopoverAppendToElement.js) directive since this component automatically appends the popover to `#bootstrap-theme`
* We can simplify the [Activities Calendar](https://github.com/compucorp/uk.co.compucorp.civicase/blob/develop/ang/civicase/ActivitiesCalendar.js) and remove a lot of the code that is associated with position the popover. This will keep things more manageable for both the Activities Calendar, and the Popover. 

### Usage

**Normal usage with a toggle button element:**
```html
<civicase-popover>
  <civicase-popover-toggle-button>Click Here!</civicase-popover-toggle-button>
  <civicase-popover-content>And you'll see this!</civicase-popover-content>
</civicase-popover>
```

**Opening the popover programmatically on top of an element:**
```html
<civicase-popover
  is-open="popoverIsOpen"
  position-reference="'#custom-trigger-button'">
  <civicase-popover-content>Here goes the content</civicase-popover-content>
</civicase-popover>

<button
  id="custom-trigger-button"
  ng-click="popoverIsOpen = true">
  Open the popover
</button>
```

You can pass the following parameters to the popover:

* **append-to** (*string|element*) where should the popover be appended to when opened. Defaults to `#bootstrap-theme`.
* **is-open** (*boolean*) can be used to programmatically  show or hide the popover.
* **popover-class** (*string*) the class that the popover will have. Important: using `class="something"` won't work on the popover since it will be attached outside of the component when opened.
* **position-reference** (*string|element*) if provided, it will open the popover on top of this element, otherwise it will open on top of the toggle button element.
* **trigger-event** (*string*) is the event that will trigger the showing and hiding of the popover. Defaults to `click`, but you can `hover` too.

Events:
* **on-open** is emitted right before opening the popover.

### The code

On init, the component will watch for `isOpen` and `positionReference` changes in order to access the DOM elements and show or hide the popover:

```js
$scope.$watch('isOpen', repositionPopover);
$scope.$watch('positionReference', repositionPopover);
```

The most important part is the `repositionPopover` function:

```js
function repositionPopover () {
  var arrowPosition, popoverDirection, position, positionReference;

  if (!$scope.isOpen) {
    return;
  }

  initPopoverReference();

  $scope.$$postDigest(function () {
    positionReference = $scope.positionReference || $toggleButton;
    position = getPopoverPositionUnderElement(positionReference);
    popoverDirection = getPopoverDirection(position);
    arrowPosition = ARROW_POSITION_VALUES[popoverDirection]
      .replace('%width%', $popoverArrow.outerWidth());

    if (popoverDirection !== 'bottom') {
      position = getPopoverPositionUnderElement(positionReference, popoverDirection);
    }

    $popoverArrow.css('left', arrowPosition);
    $popover.css(position);
  });
}
```

`initPopoverReference` will setup a direct reference to `.popover` element, append it to the `#bootstrap-theme` element, and attach hover events to it if necessary.

`$scope.$$postDigest` is an internal angular function, but it's needed to wait for the popover content to render so we can properly determine the popover width and position it accordingly. `$timeout` does not work because it is triggered more slowly and produces a noticeable jump when moving the popover.

`getPopoverPositionUnderElement` returns the position the popover will have relative to the toggle button (or the position reference)

`getPopoverDirection` will return the direction the popover should be facing in case it gets hidden by the window container. The possible values are `bottom-left`, `bottom-right`, and `bottom`. `bottom` is returned when the popover can be properly displayed without issues under the given element.

```js
ARROW_POSITION_VALUES[popoverDirection]
      .replace('%width%', $popoverArrow.outerWidth());
```
returns the position the arrow should be aligned depending on the popover position. These are the 3 possible outcomes if we consider that the popover arrow width is 22px:

* **bottom**: `''` no alignment.
* **bottom-left**: `22px` 22px from the start of the popover
* **bottom-right**: `calc(100% - 22px)` 22px from the end of the popover

if the popover can't be properly displayed inside of the window (`popoverDirection !== 'bottom'`), we need to reposition it taking into consideration the direction that will allow the popover to be displayed properly: 
```js
position = getPopoverPositionUnderElement(positionReference, popoverDirection);
```

Finally, we assign the arrow and popover positions:

```js
$popoverArrow.css('left', arrowPosition);
$popover.css(position);
```

---

`getPopoverPositionUnderElement` also has some considerations:

```js
function getPopoverPositionUnderElement ($element, direction) {
  var arrowPositionModifier, position, bootstrapThemeContainerOffset;
  direction = direction || 'bottom';
  position = $uibPosition.positionElements($element, $popover, direction, true);
  bootstrapThemeContainerOffset = $bootstrapThemeContainer.offset();
  arrowPositionModifier = getArrowPositionModifier(direction);

  return {
    top: position.top - bootstrapThemeContainerOffset.top,
    left: position.left - bootstrapThemeContainerOffset.left + arrowPositionModifier
  };
}

function getArrowPositionModifier (direction) {
  if (direction === 'bottom-left') {
    return $popoverArrow.outerWidth() / 2 * -1;
  } else if (direction === 'bottom-right') {
    return $popoverArrow.outerWidth() / 2;
  } else {
    return 0;
  }
}
```

We use `$uibPosition.positionElements` by giving it the `$element` where the popover should be displayed under, the `$popover` itself, the `direction` (defaults to `bottom`), and passing `true` so it uses the `body` as the container reference, it will gives the coordinates needed to position the `popover`.

Since we are attaching the popover to the `#bootstrap-theme` element and it has a position different than that of the `body` we need to subtract this difference to properly align the popover:
container-offset
In this case, the `240px` is the value that the `$uibPosition.positionElements` method gives us for the top position, but since the popover is attached to `#bootstrap-theme` we first need to find out what's the offset from the top of the body, in this case `140px`. That means that the popover should be positioned `100px` on the top.

### Additional contacts popover

The additional contacts popover did not have the right styles applied to it. This was fixed so it follows the invision design.

#### Before
pr-before

#### After
pr-after

### CODE REVIEW

**Comment by swastikpareek** [comment]
@reneolivo 
Initially on 
#### C51-291:Making the modal to toggle on click
`https://github.com/compucorp/uk.co.compucorp.civicase/pull/124/commits/277419231221783b37811fc4d73a3fdf003daf67`

* I attempted to try the `outsideClick` event - for the contact details popover to open on click and close when hit outside the the popover. Tried the solution mentioned in https://github.com/angular-ui/bootstrap/tree/master/src/popover#triggers. 
So , Initially I thought that using event-trigger `outsideClick` will solve our problem and we will be good to go.
I changed the css classes to be more generic and reusbale so I changed `civicase__contact-additional__popover` to `civicase__tooltip-popup-list--additional-contacts` (https://github.com/compucorp/uk.co.compucorp.civicase/pull/124/commits/277419231221783b37811fc4d73a3fdf003daf67#diff-f20fb694fca5c66b8b48c2b06ccb9648R37)

* Added a new generic styles for `civicase__tooltip-popup-list` and updated the classes for the tooltip used at other places (like https://github.com/compucorp/uk.co.compucorp.civicase/pull/124/commits/277419231221783b37811fc4d73a3fdf003daf67#diff-8ee0e472c95780f3ed52d7535e1ac1ecR42, https://github.com/compucorp/uk.co.compucorp.civicase/pull/124/commits/277419231221783b37811fc4d73a3fdf003daf67#diff-8ee0e472c95780f3ed52d7535e1ac1ecR20)

```sass
.civicase__tooltip-popup-list {
  border: 1px solid $gray-light;
  border-radius: 2px;
  box-shadow: $box-shadow-tooltip;
  color: $gray-darker;
  padding: 8px 12px;
   .popover-content {
    padding: 0;
  }
   .arrow {
    border-bottom-color: $gray-light;
  }
}
```

* Also changed `initMainContact()` to use `$event` object in `initMainContact($event)`, so that when user click on the contact icon on the case card it doesn't trigger the case card click event.

https://github.com/compucorp/uk.co.compucorp.civicase/pull/124/commits/277419231221783b37811fc4d73a3fdf003daf67#diff-7adb2c952cb4f41e7a114d941a78b361R11
https://github.com/compucorp/uk.co.compucorp.civicase/pull/124/commits/277419231221783b37811fc4d73a3fdf003daf67#diff-95760ad4747ea1c72303b931cebb4a56R34


### C51-291: partial work on the popup

I later checked that there is a issue with using `civicaseAppendPopoverToElement` directive, as clicking/hovering over other popup element when there is already a popover element appended to `#bootstrap-element` using `civicaseAppendPopoverToElement`, realigns itself w.r.t to the `body` element and gets wring position. After spending some time on investigation I figured that we anyways need to rebuild the popup to fix it forever, 

So, I took your technical solution mentioned in C51-291. Creating a new popup directive from scratch and tried to setup/trigger different events in it (e.g. click, mouseover, etc.)

* `ang/civicase/Popover.html`

```html
<div class="civicase__popover">
    <span ng-transclude="toggleButton"></span>
    <div class="popover bottom civicase__popover-box" style="display: block" ng-if="isOpen">
      <div class="arrow"></div>
      <div class="popover-content" ng-transclude="content">
      </div>
    </div>
  </div>
// Referred from the tech details of C51-291
```
*  in `ang/civicase/Popover.js`
I tried extending the solution you mentioned in C51-291 technical details and tried implementing the event listeners to the directive.
```js
$scope.triggerEvent = $scope.triggerEvent || 'click';
...
...

function attachEventListeners () { // function to add/attach event listeners passed as an attribute to the directive
  $toggleButton.on($scope.triggerEvent, function (event) {
    if (!$scope.isOpen) {
      $rootScope.$broadcast('civicase::popover::close-all');
    }
     $scope.togglePopoverState();
    $scope.$apply();
    event.stopPropagation();
  });
   $scope.$on('civicase::popover::close-all', function () {
    $scope.isOpen = false;
  });
   if (!$('body').hasClass('civicase__popup-attached')) { // attaching `Close popup when clicked outside` event 
    $(document).on('click', function ($event) {
      if ($('.civicase__popover-box').find($event.target).length === 0) {
        $rootScope.$broadcast('civicase::popover::close-all'); // closing all other popups too 
      }
    });
    $('body').addClass('civicase__popup-attached');
  }
}
```

Other JS code was implemented from the technical steps. And the Css code is added because I was trying to implemnent the solution w/o using `civicaseAppendPopoverToElement` Directive, and was appending it to body element so need to wrap the css code written in the previous PR inside `.___outside-namespace` css selector.
Also did some other UX improvements/regression my code brough up 
```css
  padding-left: 0 !important;
  padding-right: 0 !important;
```

However, I don't think we need `$popover.show()` in `ang/civicase/PopoverAppendToElement.js`  https://github.com/compucorp/uk.co.compucorp.civicase/pull/124/commits/05fe80e17936ee20f53b7437d19f020b0059abd7#diff-5d4e6bbacd36965737d4ad8139854419R27. We can get rid of it.

##########################################################################
These were the changes and my thought process while working on C51-291. 


 **PS: While working on it I bumped up on an technical issue and thought of asking debarshi for the solution. (the same day Debarshi joined back from his holidays) and he suggested me to work on some other more priority issue and suggested to talk to you on the solution ( I was implementing ) for this ticket before working on it further. Also, he suggested me to assign it to you as you have more idea and experience of the problem (which resulted in the call between you me and debarshi - on the day you joined back from holidays).**

Please let me know if there is any confusion or more details needed on the commits/work I done on C51-291.

Thanks 


cc @deb1990

File: ang/civicase/Popover.js [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+      })();
+
+      /**
+       * Switch between open/ closed state
+       */
+      $scope.togglePopoverState = function () {
+        $scope.isOpen = !$scope.isOpen;
+      };
+
+      function attachEventListeners () {
```
Comment (swastikpareek) [inline ang/civicase/Popover.js]: Docs missing.

File: ang/civicase/Popover.js [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      function repositionPopover () {
+        var arrowPosition, popoverDirection, position, positionReference;
+
+        if (!$scope.isOpen) {
+          return;
+        }
+
+        initPopoverReference();
+
+        /**
```
Comment (swastikpareek) [inline ang/civicase/Popover.js]: Good catch. Thanks for the detailed info/reason.

File: ang/civicase/ContactCard.html [hunk 3]
```diff
@@ -126,4 +127,5 @@ <h2>{{mainContact.display_name}}</h2>
       </div>
     </div>
   </script>
+
```
Comment (swastikpareek) [inline ang/civicase/ContactCard.html]: Can we remove this extra line.

File: ang/civicase/Popover.html [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<div class="civicase__popover"
+  aria-haspopup="true"
+  aria-expanded="{{isOpen}}">
+  <span ng-transclude="toggleButton"></span>
+  <div class="popover bottom civicase__popover-box {{popoverClass}}" ng-if="isOpen">
```
Comment (swastikpareek) [inline ang/civicase/Popover.html]: Can we implement `class="popover bottom civicase__popover-box {{popoverClass}}"`  using `ng-class` ?

**Comment by reneolivo** [comment]
As discussed with @swastikpareek the additionals contacts dimensions will be fixed in a different ticket as to avoid bloating the scope of this one.

**Pull request merged**